### DISCUSSION TITLE
Why git lfs binary files are converted to pointers during git checkout?

### DISCUSSION THREAD

**Discussion by withtimesgo1115** [git-lfs/git-lfs#5716] [Q&A]
### Why git lfs binary files are converted to pointers during git checkout?

*Describe your question in as much detail as possible:*

In our ci job, the GIT_LFS_SKIP_SMUDGE is set to 0 because we need to use the newest git lfs binary files. We want to speed up the CI job and the idea is to use the pvc mounted to each k8s job pod for consistent storage, to be more specific, to reuse the git repo and the bazel cache. However, the git lfs pull process is too slow(around 5min) :frowning: . It is unexpected because we suppose the ci job can use the git repo instead of pull everything each time.
To debug this problem, I did a lot of work:
I added some git configs to the runner to trace the git operation and performance like the following:
GIT_TRACE: 1
GIT_TRACE_PERFORMANCE: 1
GIT_TRACE_SETUP: 1
Meanwhile, I also added "ls -la /path" logs before and after the critical operation (git fetch/git checkout/git clean) in the Gitlab Runner source and build the binaries for gitlab runner.

![image](https://github.com/git-lfs/git-lfs/assets/56740244/799c3a93-b40a-4e00-bb83-88b03e6004d1)

![image](https://github.com/git-lfs/git-lfs/assets/56740244/938fd3c6-eaaf-4208-bcef-aa522e660145)

![image](https://github.com/git-lfs/git-lfs/assets/56740244/594c3892-9021-45f2-95c2-7e07e7a8ead1)

You can see the lfs binaries in the pvc are converted to pointers during check out process! This is the reason why the following git lfs pull process is slow. 
![image](https://github.com/git-lfs/git-lfs/assets/56740244/4b810ea9-bc55-4e81-b538-e9bd3a60000c)

**The question is that why git lfs binary files are converted to pointers during git checkout? Anyone can help me out??? Thank you!!!** 

### Steps to reproduce 

- create pvc/pv resource
- use pvc working for consistent storage for our project git repo and bazel cache.
- deploy the k8s runner by helm chart 
- trigger a pipeline and the job is executed by the pvc runner

Versions 
- git lfs 3.2.0
- Gitlab 16.8.1

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.  Would you be able to provide the exact series of commands you are running, along with the relevant environment variables?  For instance, something like:

```
$ GIT_LFS_SKIP_SMUDGE=0 git clone https://...
$ cd ...
$ GIT_LFS_SKIP_SMUDGE=0 git checkout ...
```

or whatever the specific, complete set of steps is for your case?  We don't need private information, of course.

**Reply by withtimesgo1115** [reply]
Hi, Chris, Thank you for your reply! 
Honestly, I cannot reproduce in the local environment. 
The Commands I used are like this:
```
export GIT_TRACE: 1
export GIT_TRACE_PERFORMANCE: 1
export GIT_TRACE_SETUP: 1
export GIT_LFS_SKIP_SMUDGE=1
ls -la /lfs-files-path    // binaries here
git fetch origin +refs/pipelines/622126:refs/pipelines/622126 +refs/tags/proj-dev_jerry_lfs_test_202404181449-20240418_144911:refs/tags/proj-dev_jerry_lfs_test_202404181449-20240418_144911 --depth 10 --prune --quiet
ls -la /lfs-files-path    // binaries here
git checkout -f -q 47bc63d0d63c928a17123c748d26fc390528a763
ls -la /lfs-files-path    // pointers here
git clean -ffdx
ls -la /lfs-files-path    // pointers here
```
As you can see from the logs, some low-level commands are executed during checkout, and I am unclear about what happened. 
```
08:21:50.534261 git.c:439               trace: built-in: git checkout -f -q 47bc63d0d63c928a17123c748d26fc390528a763
08:21:50.573562 read-cache.c:2306       performance: 0.038782030 s:  read cache .git/index
08:21:50.582862 preload-index.c:147     performance: 0.009259375 s:  preload index
08:21:50.737844 unpack-trees.c:1597     performance: 0.154274122 s:   traverse_trees
08:21:52.119806 run-command.c:663       trace: run_command: 'git-lfs filter-process'
08:21:52.196322 trace git-lfs: exec: git 'version'
08:21:52.199737 trace git-lfs: exec: git 'config' '-l'
08:21:52.201169 trace git-lfs: Install hook: pre-push, force=false, path=/tmp/repo/project/.git/hooks/pre-push, upgrading...
08:21:52.201610 trace git-lfs: Install hook: post-checkout, force=false, path=/tmp/repo/project/.git/hooks/post-checkout
08:21:52.201652 trace git-lfs: Install hook: post-commit, force=false, path=/tmp/repo/project/.git/hooks/post-commit, upgrading...
08:21:52.201923 trace git-lfs: Install hook: post-merge, force=false, path=/tmp/repo/project/.git/hooks/post-merge, upgrading...
08:21:52.202267 trace git-lfs: Initialize filter-process
08:21:52.202421 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' 'HEAD' '--symbolic-full-name' 'HEAD'
08:21:52.204091 trace git-lfs: tq: running as batched queue, batch size of 100
08:22:03.728526 unpack-trees.c:452      performance: 12.990206817 s:   check_updates
08:22:03.750622 cache-tree.c:446        performance: 0.003904941 s:   cache_tree_update
08:22:03.750941 unpack-trees.c:1696     performance: 13.167988622 s:  unpack_trees
08:22:03.808315 read-cache.c:3033       performance: 0.053454216 s:  write index, changed mask = 2a
08:22:03.811172 run-command.c:663       trace: run_command: .git/hooks/post-checkout 47bc63d0d63c928a17123c748d26fc390528a763 47bc63d0d63c928a17123c748d26fc390528a763 1
08:22:03.815847 git.c:703               trace: exec: git-lfs post-checkout 47bc63d0d63c928a17123c748d26fc390528a763 47bc63d0d63c928a17123c748d26fc390528a763 1
08:22:03.815877 run-command.c:663       trace: run_command: git-lfs post-checkout 47bc63d0d63c928a17123c748d26fc390528a763 47bc63d0d63c928a17123c748d26fc390528a763 1
08:22:03.825591 trace git-lfs: exec: git 'version'
08:22:03.828558 trace git-lfs: exec: git 'config' '-l'
08:22:03.830007 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' 'HEAD' '--symbolic-full-name' 'HEAD'
08:22:03.831630 trace git-lfs: NewLsFiles: running in /tmp/repo/project git ls-files -z --others --cached --exclude-standard
08:22:04.209161 run-command.c:43        trace: run_command: running exit handler for pid 53
08:22:04.209271 trace git-lfs: filepathfilter: rewrite ".git" as "**/.git/**"
08:22:04.209297 trace git-lfs: filepathfilter: rewrite "**/.git" as "**/.git"
08:22:04.209367 trace git-lfs: filepathfilter: accepting "tmp"
08:22:04.211815 trace.c:475             performance: 13.677803982 s: git command: git checkout -f -q 47bc63d0d63c928a17123c748d26fc390528a763
```
Another puzzling phenomenon I met is that the conversion between binaries and pointers sometimes wouldn't happen for gitlab runner docker executor but every time for gitlab runner kubernetes executor.

**Reply by chrisd8088** [reply]
Hmm, interesting.  One thing I can say is that there's a difference between `GIT_LFS_SKIP_SMUDGE=1` and `GIT_LFS_SKIP_SMUDGE=0`.  The latter (`=0`) means "convert pointers to objects" and the former (`=1`) means _don't_ convert pointers to objects.

So if you're mixing those two settings (you mention "`GIT_LFS_SKIP_SMUDGE` is set to `0`" in the original question, and `export GIT_LFS_SKIP_SMUDGE=1` above), you might get variable results.

**Reply by withtimesgo1115** [reply]
Actually, I set the GIT_LFS_SKIP_SMUDGE = 0 for my Gitlab CI Job but I found this value would be forced converted to 1 in the Gitlab Runner source code.
```
func (b *AbstractShell) writeCloneFetchCmds(w ShellWriter, info common.ShellScriptInfo) error {
	build := info.Build

	// If LFS smudging was disabled by the user (by setting the GIT_LFS_SKIP_SMUDGE variable
	// when defining the job) we're skipping this step.
	//
	// In other case we're disabling smudging here to prevent us from memory
	// allocation failures.
	//
	// Please read https://gitlab.com/gitlab-org/gitlab-runner/issues/3366 and
	// https://github.com/git-lfs/git-lfs/issues/3524 for context.
	if !build.IsLFSSmudgeDisabled() {
		w.Variable(common.JobVariable{Key: "GIT_LFS_SKIP_SMUDGE", Value: "1"})
	}

	err := b.handleGetSourcesStrategy(w, build)
	if err != nil {
		return err
	}

	if build.GetGitCheckout() {
		b.writeCheckoutCmd(w, build)

		// If LFS smudging was disabled by the user (by setting the GIT_LFS_SKIP_SMUDGE variable
		// when defining the job) we're skipping this step.
		//
		// In other case, because we've disabled LFS smudging above, we need now manually call
		// `git lfs pull` to fetch and checkout all LFS objects that may be present in
		// the repository.
		//
		// Repositories without LFS objects (and without any LFS metadata) will be not
		// affected by this command.
		//
		// Please read https://gitlab.com/gitlab-org/gitlab-runner/issues/3366 and
		// https://github.com/git-lfs/git-lfs/issues/3524 for context.
		if !build.IsLFSSmudgeDisabled() {
			w.IfCmd("git", "lfs", "version")
			w.Command("git", "lfs", "pull")
			w.EmptyLine()
			w.EndIf()
		}
	} else {
		w.Noticef("Skipping Git checkout")
	}

	return nil
}
```

**Reply by withtimesgo1115** [reply]
@bk2204 @chrisd8088 any suggestions? I also used strace to see what happened where I found that there are many unlink and write system calls during checkout process.

**Reply by chrisd8088** [reply]
I'm sorry, we're not too familiar with GitLab, so you might need to contact their Support team to find out what their code is doing.  But if it's setting `GIT_LFS_SKIP_SMUDGE=1` then it seems like it could be the source of the problem.

If you're able to reproduce this in a local environment, please do share the reproduction steps!  Until then I fear it will be a bit difficult to analyze.

**Reply by bk2204** [reply]
I agree with Chris that the CI is the source of the problem here.  It probably should not be forcing `GIT_LFS_SKIP_SMUDGE` to 1, and if it is, that's why you're seeing that.  You'll probably want to open a bug with them asking them to stop that behaviour or allow it to be overridden.

In the mean time, you can get the large files by running `git lfs pull`.
